### DISCUSSION TITLE
Having issues mapping between same 2 types.

### DISCUSSION THREAD

**Discussion by OTonGitHub** [riok/mapperly#1281] [General]
I have a domain entity called Activity, this is being returned by my API. When I get it, I want to update the database Activity.. So I did this,

![image](https://github.com/riok/mapperly/assets/81725875/1df004aa-f1ec-4826-afbf-d663beabe0e4)

Generated File looks like this.

![image](https://github.com/riok/mapperly/assets/81725875/bf355d7b-de2e-4016-ad8c-da0b8d34ba37)

and my mapping code looks like this and has this warning..

![image](https://github.com/riok/mapperly/assets/81725875/93ccfdec-cd91-442f-85a1-21a1d415348e)

and the mapping simply doesnt work, nothing changed.. I know I am probably doing a really stupid mistake, I love the idea of this mapper and am just trying to figure this out but I couldn't

**Comment by OTonGitHub** [comment]
I am so stupid, I forgot this would need me to inform EF core to track update changes ehhe.. Perks of using an object mapper for the first time, sorry.

**Comment by OTonGitHub** [comment]
Actually this didn't still solve the problem for me.. this is getting embarrassing.

**Comment by latonz** [comment]
With the current implementation Mapperly just returns your input argument as you can see in your implementation screenshot (`return activity`). What happens is:
1. you use the id of the request‘s activity and fetch the stored activity from the DB
2. you store the activity of the request in your `activity` variable. By doing this you overwrite the reference to the activity you fetched from the DB.
3. You store the changes in EF. But there are no changes as you never updated the object you fetched from the DB.

Is the request body of the same type as your entity? Do you want to replace the object or just update some properties?

Depending on the questions above you have several options to fix your code:
1. use an existing target mapping with your database object as the target and update it with the data of the request: [docs](https://mapperly.riok.app/docs/configuration/existing-target/).
2. use your request object directly and attach it to the db context and mark it as updated (if you are using the same type for the request body as for your database entity and want to replace the object completely, using the same type has pros/cons and is a different discussable topic but if you opt for this no mapper is needed at all)
3. Map your request object to the entity type, attach and mark as updated

**Reply by OTonGitHub** [reply]
Thanks for the reply taking your time to help resolve my issue.

I understand that, not using a DTO on my request body may not be the "best practice", it's just the case for now and I am willing to discuss more on it in another thread.

But regarding my issue, it was my fault for not checking the docs as well as I have to, but I have taken some time to go through the documentation and the advice you gave, and have made the following changes to my mapping class

![image](https://github.com/riok/mapperly/assets/81725875/005be5c7-d43c-4d5f-b43e-d1bfa8df9279)

And then I nested `<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>` in my project file and built my project, the generated source file looks good, it adds the necessary null checks. 

- side note: *when I right click on project and build, it does-not seems to update the generated source file, I am using vscode, so I had to use the dotnet build command in the terminal for it to work.*

But when mapping, I notice this undesired behavior, here is the object I am about to update, in it's original state.

![image](https://github.com/riok/mapperly/assets/81725875/d6829e1d-89f0-49b1-9990-382ba1a9f2c2)

This is my update request sent to the API.

![image](https://github.com/riok/mapperly/assets/81725875/8cc29773-c63d-4f0a-9537-8be9204841f6)

And this is my API

![image](https://github.com/riok/mapperly/assets/81725875/36428c87-06ab-41ec-954b-a48dabe96e2b)

And after updating the object now, this is what it looks like.

![image](https://github.com/riok/mapperly/assets/81725875/c7ebc19d-517c-44d2-8cdd-fc363f1b0608)

Which is pretty much what I expected except the DateTime field. I set the property like this in my model.
```public DateTime Date { get; set; }``` I understand DateTime is not gonna be nullable in the database.

But it is not provided in the incoming object, so why is Mapperly trying to map it?

![image](https://github.com/riok/mapperly/assets/81725875/de4448f5-0298-419e-b5fa-dc414eca2be5)

From my understanding of the code (I am still a beginner) it looks to me like that, when the API accepts the request, if DateTime is not set, it's then set to a default value when the object is created?

Or I am understanding this wrong, in this case, how can I solve this? I still need DateTime to not be a nullable in the database, but I just want to update the properties that are included in the request body only.
If I am understanding the problem correctly, one way to solve this would be to set the DateTime of the source object to what's already available in the database before providing it to mapperly. But how would I know the user did not include it in the request body if API sets it to a default value when accepting the object.

I apologize if my issue here is too begginer-ly. Any little guidance or help is appreciated. Thank you.

**Reply by latonz** [reply]
I'm sorry, but my time is limited and I am unable to provide further assistance in this case.
Mapperly seems to be working as intended. For more information, please consult the documentation.

**Reply by OTonGitHub** [reply]
That's fine, and thank you, I should have closed this earlier.
